### PR TITLE
FIX + feat: add MessageManager layer for centralized history coordination

### PR DESCRIPTION
**Pull request by hannesrudolph**
## Summary

Implements the proposal from `docs/proposals/message-manager-layer.md`.

This introduces a **MessageManager** class that centralizes all conversation rewind operations, ensuring consistent handling of both UI (`clineMessages`) and API (`apiConversationHistory`) message histories.

## Problem Solved

The codebase previously had **two separate code paths** for removing messages from conversation history:
1. When restoring a checkpoint from **before** a context-management event, the API history is now properly rehydrated. Previously, messages tagged with `condenseParent` or `truncationParent` would remain hidden even though their parent Summary/marker no longer existed. Now, `cleanupAfterTruncation()` properly removes these orphaned tags, making the condensed/truncated messages visible again.
2. Regular delete/edit operations via `removeMessagesThisAndSubsequent()` in `webviewMessageHandler.ts`
3. Checkpoint restore operations via `checkpointRestore()` in `checkpoints/index.ts`

These paths handled message removal differently, leading to potential inconsistencies in how context-management events (condensing, truncation) were processed.

## Key Bug Fix

Checkpoint restore now properly handles context-management events:
- ✅ Collects `condenseId`s from removed `condense_context` messages
- ✅ Collects `truncationId`s from removed `sliding_window_truncation` messages
- ✅ Removes orphaned Summary messages from API history
- ✅ Removes orphaned truncation markers from API history
- ✅ Calls `cleanupAfterTruncation()` to clear orphaned parent tags

### New Files
| File | Description |
|------|-------------|
| `src/core/message-manager/index.ts` | MessageManager class with `rewindToTimestamp()`, `rewindToIndex()`, and internal methods |
| `src/core/message-manager/index.spec.ts` | 22 comprehensive unit tests |

### Modified Files
| File | Changes |
|------|---------|
| `src/core/webview/webviewMessageHandler.ts` | Replaced 79-line inline `removeMessagesThisAndSubsequent()` with MessageManager calls |
| `src/core/checkpoints/index.ts` | Refactored to use MessageManager (the critical bug fix) |
| `src/core/task/Task.ts` | Added `messageManager` getter with lazy initialization |
| `src/core/webview/__tests__/webviewMessageHandler.delete.spec.ts` | Updated 1 test for timestamp-based behavior |

## Testing

- All **4508 tests pass** with no regressions
- 22 new MessageManager unit tests cover:
  - Basic rewind operations
  - Condense handling (preservation and removal)
  - Truncation handling (preservation and removal)
  - Checkpoint scenarios (the key bug fix)
  - Skip cleanup option
  - Combined scenarios and edge cases

## Architecture

```
┌─────────────────────────────────────────────────────────────────────┐
│                        Message Operations                           │
├─────────────────┬─────────────────┬─────────────────────────────────┤
│  Delete Message │  Edit Message   │  Checkpoint Restore             │
│  (webview)      │  (webview)      │  (checkpoints)                  │
└────────┬────────┴────────┬────────┴───────────────┬─────────────────┘
         │                 │                        │
         └─────────────────┼────────────────────────┘
                           │
                           ▼
              ┌────────────────────────┐
              │    MessageManager      │
              │                        │
              │  • rewindToTimestamp() │
              │  • rewindToIndex()     │
              └───────────┬────────────┘
                          │
        ┌─────────────────┼─────────────────┐
        │                 │                 │
        ▼                 ▼                 ▼
┌───────────────┐ ┌───────────────┐ ┌───────────────┐
│ Collect IDs   │ │ Truncate      │ │ Cleanup       │
│ from removed  │ │ clineMessages │ │ orphaned      │
│ context events│ │ & apiHistory  │ │ tags          │
└───────────────┘ └───────────────┘ └───────────────┘
```

----

> [!IMPORTANT]
> Introduces `MessageManager` for centralized message history operations, ensuring consistent handling across UI and API paths, with comprehensive testing.
> 
>   - **Behavior**:
>     - Introduces `MessageManager` class in `src/core/message-manager/index.ts` for centralized message history operations.
>     - Replaces inline message removal logic in `webviewMessageHandler.ts` with `MessageManager` calls.
>     - Ensures consistent handling of message histories across UI and API paths.
>   - **Bug Fix**:
>     - Fixes checkpoint restore handling of context-management events by removing orphaned tags and markers.
>   - **Testing**:
>     - Adds `src/core/message-manager/index.spec.ts` with 22 unit tests covering rewind operations, condense and truncation handling, and edge cases.
>   - **Misc**:
>     - Updates `Task.ts` to include a `messageManager` getter for lazy initialization.
>     - Minor test updates in `webviewMessageHandler.delete.spec.ts` for timestamp-based behavior.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for cef2492e699ebefd1101abd9670d980ff81ff3c0. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by ghost** [comment]
> **Oroocle**  Clock &nbsp;&nbsp;[Follow along on Roo Cloud](https://app.roocode.com/cloud-jobs/Z33PtFqt?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review.sync)

Review completed with one suggestion about consolidating MessageManager usage through the Task.messageManager getter for consistency and future extensibility.

- [x] Route all MessageManager usages through Task.messageManager to avoid multiple instances and keep a single coordination point.

> <details>
> <summary>Previous reviews</summary>
> 
> - cef2492: [Review #1](https://app.roocode.com/cloud-jobs/mKKenV2L?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - 39ca0ba: [Review #2](https://app.roocode.com/cloud-jobs/s00TjA0d?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - 956b3a2: [Review #3](https://app.roocode.com/cloud-jobs/fRRjsFuG?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - b2a8b40: [Review #4](https://app.roocode.com/cloud-jobs/fRCfsFuG?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - 7b0c108: [Review #5](https://app.roocode.com/cloud-jobs/BcszlA3W?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - b8c3728: [Review #6](https://app.roocode.com/cloud-jobs/3dKLF6XD?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> - a3ee9bf: [Review #7](https://app.roocode.com/cloud-jobs/BcsZlA3W?utm_source=cloud-job-url&utm_medium=link&utm_campaign=previous-reviews)
> </details>
<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=9842&commentId=3614370416).</sub>

File: src/core/task/Task.ts [hunk 1]
```diff
@@ -4032,6 +4036,24 @@ export class Task extends EventEmitter<TaskEvents> implements TaskLike {
...
+	 * Uses lazy initialization - the MessageManager is only created when first accessed.
+	 * Subsequent accesses return the same cached instance.
+	 *
+	 * @example
+	 * ```typescript
+	 * // Rewind conversation to a specific timestamp
+	 * await task.messageManager.rewindToTimestamp(ts)
+	 * ```
+	 */
+	get messageManager(): MessageManager {
```
Comment (ghost) [inline src/core/task/Task.ts:4061]: The new MessageManager centralizes rewind logic nicely, but it is not yet the single entry point for all rewind operations. Task.checkpointRestore still constructs a new MessageManager directly instead of going through the Task.messageManager getter, and any future direct instantiations would also bypass the shared instance. To keep behavior consistent and make it easier to evolve MessageManager (e.g., adding internal state or instrumentation), consider routing all usage through the Task.messageManager getter and updating the class docs accordingly.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=9842&fixId=f809a01819) or mention @roomote and request a fix.</sub>

Comment (hannesrudolph) [reply src/core/task/Task.ts:4061]: @roomote without making any changes, can you tell me if this comment has been addressed by followup PRs already?

Comment (ghost) [reply src/core/task/Task.ts:4061]: > **Fixaroo**  Clock &nbsp;&nbsp;[See task on Roo Cloud](https://app.roocode.com/cloud-jobs/LVxMOCiy?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.comment.respond)

**Yes, this comment has been addressed by followup commits in this PR.**

The concern about routing all MessageManager usages through `Task.messageManager` was addressed in two subsequent commits:

1. **Commit b8c3728** ("refactor: consolidate MessageManager usage through Task.messageManager getter"):
   - `checkpoints/index.ts`: Changed from `new MessageManager(task)` to `task.messageManager`
   - `webviewMessageHandler.ts`: Changed from `new MessageManager(currentCline)` to `currentCline.messageManager`

2. **Commit b2a8b40** ("docs: add architectural guidance for MessageManager getter usage"):
   - Updated the `Task.messageManager` getter documentation to explicitly state that all MessageManager operations must go through the getter rather than instantiating new instances directly

The current state of the PR now routes all MessageManager usage through the `Task.messageManager` getter as recommended.

File: src/core/checkpoints/index.ts [hunk 2]
```diff
@@ -258,20 +258,20 @@ export async function checkpointRestore(
...
 
-			// For delete operations, exclude the checkpoint message itself
-			// For edit operations, include the checkpoint message (to be edited)
-			const endIndex = operation === "edit" ? index + 1 : index
-			await task.overwriteClineMessages(task.clineMessages.slice(0, endIndex))
+			// Use MessageManager to properly handle context-management events
+			// This ensures orphaned Summary messages and truncation markers are cleaned up
+			await task.messageManager.rewindToTimestamp(ts, {
+				includeTargetMessage: operation === "edit",
+			})
```
Comment (mrubens) [inline src/core/checkpoints/index.ts:272]: Not a blocker, but I continue to think that it would be nice if we didn't need to special case edit in the way that we have been doing

Comment (hannesrudolph) [reply src/core/checkpoints/index.ts:272]: Tomorrow's problem?
